alg_convex_hull.go

```diff
@@ -1,110 +1,191 @@
+}
+
+// grahamScan returns the convex hull of the input points. It will either
+// represent the empty set (zero points), a point (one point), a line (2
+// points), or a closed polygon (>= 3 points).
 func grahamScan(ps []XY) []XY {
-	if len(ps) < 3 {
-		return nil
+	if len(ps) <= 1 {
+		return ps
 	}
 
 	sortByPolarAngle(ps)
 
-	resultStack := make([]XY, 2, len(ps))
-	copy(resultStack, ps[:2])
-	toDoStack := make([]XY, len(ps)-2)
-	copy(toDoStack, ps[2:])
+	// Append the lowest-then-leftmost point so that the polygon will be closed.
+	ps = append(ps, ps[0])
 
-	for i := 0; i < len(toDoStack); i++ {
-		ori := orientation(resultStack[len(resultStack)-2], resultStack[len(resultStack)-1], toDoStack[i])
-		switch {
-		case ori == leftTurn:
-			resultStack = append(resultStack, toDoStack[i])
-		default:
-			resultStack = resultStack[:len(resultStack)-1]
-			resultStack = append(resultStack, toDoStack[i])
+	var i int // tracks progress through the ps slice
+	var stack pointStack
+	stack.push(ps[0])
+	i++
+	for i < len(ps) && len(stack) < 2 {
+		if !stack.top().Equals(ps[i]) {
+			stack.push(ps[i])
 		}
+		i++
 	}
 
-	return append(resultStack, resultStack[0])
+	for i < len(ps) {
+		ori := orientation(stack.underTop(), stack.top(), ps[i])
+		switch ori {
+		case leftTurn:
+			// This point _might_ be part of the convex hull. It will be popped
+			// later if it actually isn't part of the convex hull.
+			stack.push(ps[i])
+		case collinear:
+			// This point is part of the convex hull, so long as it extends the
+			// current line segment (in which case the preceding point is
+			// _not_ part of the convex hull).
+			if distanceSq(stack.underTop(), ps[i]).GT(distanceSq(stack.underTop(), stack.top())) {
+				stack.pop()
+				stack.push(ps[i])
+			}
+		default:
+			//  The preceding point was _not_ part of the convex hull (so it is
+			//  popped). Potentially the new point is just an extension of a
+			//  straight line (so pop the preceding point in that case so as to
+			//  eliminate collinear points).
+			stack.pop()
+			if orientation(stack.underTop(), stack.top(), ps[i]) == collinear {
+				stack.pop()
+			}
+			stack.push(ps[i])
+		}
+		i++
+	}
+	return stack
 }
 
-// soryByPolarAngle sorts the points by their polar angle
+// sortByPolarAngle sorts the points by their polar angle relative to the
+// lowest-then-leftmost anchor point.
 func sortByPolarAngle(ps []XY) {
+	// the lowest-then-leftmost (anchor) point comes first
 	ltlp := ltl(ps)
-
-	// swap the ltl point with first point
 	ps[ltlp], ps[0] = ps[0], ps[ltlp]
+	anchor := ps[0]
 
-	virtualPoint := ps[0]
-
+	ps = ps[1:] // only sort the remaining points
 	sort.Slice(ps, func(i, j int) bool {
-		if i == 0 {
-			return false
+		if anchor.Equals(ps[i]) {
+			return true
 		}
-		ori := orientation(virtualPoint, ps[i], ps[j])
-
-		if ori == collinear {
-			return distanceSq(virtualPoint, ps[i]).LT(distanceSq(virtualPoint, ps[j]))
+		if anchor.Equals(ps[j]) {
+			return false
```

Comment:
Don't quite understand this part and above? Would you please elaborate it?

Reply:
I've added a comment, let me know if that doesn't clear it up, and I can explain further.